### PR TITLE
feat: undo manager

### LINKED ISSUE #2: Missing Undo / Redo
Due to technical limitations undo and redo using Copilot is not possible. This makes it hard to undo, especially since files are not open in the editor - user doesn't even know what has been changed.

### PR DESCRIPTION
**Pull request by MarcinVaadin**
## Description

Undo manager. 

Undo/Redo availability notification in Copilot will be added after implementing 2way communication with plugin.

Fixes #2

## Type of change

- [ ] Bugfix
- [x] Feature

## Checklist

- [x] I have read the contribution guide: https://vaadin.com/docs/latest/guide/contributing/overview/
- [x] I have added a description following the guideline.
- [x] The issue is created in the corresponding repository and I have referenced it.
- [x] I have added tests to ensure my change is effective and works as intended.
- [x] New and existing tests are passing locally with my change.
- [x] I have performed self-review and corrected misspellings.

#### Additional for `Feature` type of change

- [x] Enhancement / new feature was discussed in a corresponding GitHub issue and Acceptance Criteria were created.

### CODE REVIEW

**Comment by abdullahtellioglu** [comment]
In general, it works however it stops if you modify the code. You cannot undo anymore

**Comment by MarcinVaadin** [comment]
> In general, it works however it stops if you modify the code. You cannot undo anymore

Yes, that by purpose not to affect user changes. Final solution after 2-way plugin communication would be to disable undo/redo on Copilot if not possible.

File: src/helpers/handlers.ts [hunk 1]
```diff
@@ -64,7 +68,13 @@ export async function writeFileHandler(data: WriteCommandData) {
...
         // save changes
         vscode.workspace.openTextDocument(uri).then(doc => {
-            doc.save();
+            undoManager.lockDocument(doc);
+            doc.save().then(result => {
+                if (result) {
+                    undoManager.pluginFileWritten(doc);
+                }
+                undoManager.unlockDocument(doc);
+            });
```
Comment (czp13) [inline src/helpers/handlers.ts:77]: Not sure in what cases this promise can fail, but would it make sense to do the unlock in the finally part?:

```
doc.save()
    .then(result => {
        if (result) {
            undoManager.pluginUndoRedoPerformed(document, operation);
        }
    })
    .catch(error => {
        // Handle any error that might occur during the save or the undo operation
        console.error("Error during save operation:", error);
    })
    .finally(() => {
        undoManager.unlockDocument(doc);
    });
```

**Comment by abdullahtellioglu** [comment]
Not sure this approach would work with formatter plugins that update files after saving.

File: src/helpers/handlers.ts [hunk 2]
```diff
@@ -81,12 +91,22 @@ export async function undoRedoHandler(data: UndoRedoCommandData, operation: 'und
...
             await vscode.commands.executeCommand(operation);
-            await vscode.workspace.save(uri);
-            await vscode.commands.executeCommand('workbench.action.closeActiveEditor');
+            vscode.workspace.openTextDocument(uri).then(doc => {
+                undoManager.lockDocument(doc);
+                doc.save().then(result => {
+                    if (result) {
+                        undoManager.pluginUndoRedoPerformed(document, operation);
+                    }
+                    undoManager.unlockDocument(doc);
```
Comment (czp13) [inline src/helpers/handlers.ts:107]: same as above just to make sure we always unlock the resource! 🙇

**Review by czp13 — APPROVED** [review]
Code seems good to me, and worked as well while tested after installed the plugin latest version.

**Comment by czp13** [comment]
> Not sure this approach would work with formatter plugins that update files after saving.

This is a good point, I could not find an option in VS Code  (maybe I was blind to see it) other than installing a separate extension for auto-formatting?  

You can always use the hotkey to format, but in that case the document is locked so the undo/redo should happen before you can use formatting hotkey.

Can we test this somehow? Or need to be tested with external plugin like [Prettier Code Formatter extension](https://geekflare.com/auto-format-in-vs-code/)?

**Comment by MarcinVaadin** [comment]
> Not sure this approach would work with formatter plugins that update files after saving.

Yes, that is the same problem we have with IntelliJ plugin, on save actions are causing undo redo issues. Not solved yet.

File: src/helpers/handlers.ts (hunk 1 above)
Comment (MarcinVaadin) [reply src/helpers/handlers.ts:77]: Yes, good tip, finally seems proper place for it 🙇

**Comment by MarcinVaadin** [comment]
Follow up https://github.com/vaadin/copilot-internal/issues/2102

**Pull request merged**